sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java

```diff
@@ -0,0 +1,318 @@
+      session = cluster.connect();
+      LOG.debug("Query: " + splitQuery);
+      resultSet = session.execute(splitQuery);
+
+      final MappingManager mappingManager = new MappingManager(session);
+      Mapper mapper = mappingManager.mapper(entity);
+      iterator = mapper.map(resultSet).iterator();
+      return advance();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      if (iterator.hasNext()) {
+        current = iterator.next();
+        return true;
+      }
+      current = null;
+      return false;
+    }
+
+    @Override
+    public void close() {
+      LOG.debug("Closing Cassandra reader");
+      if (session != null) {
+        session.close();
+      }
+      if (cluster != null) {
+        cluster.close();
+      }
+    }
+
+    @Override
+    public T getCurrent() throws NoSuchElementException {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+  }
+
+  @Override
+  public Reader<T> createReader(String query, Class<T> entity) throws IOException {
+    return new CassandraReaderImpl<>(query, entity);
+  }
+
+  @Override
+  public long getEstimatedSizeBytes() {
+    try (Cluster cluster = connectionConfiguration.getCluster()) {
+      List<TokenRange> tokenRanges = getTokenRanges(cluster, connectionConfiguration, table);
+      long size = 0L;
+      for (TokenRange tokenRange : tokenRanges) {
+        size += tokenRange.meanPartitionSize * tokenRange.partitionCount;
+      }
+      return Math.round(size / getRingFraction(cluster, tokenRanges));
+    }
+  }
+
+  @Override
+  public List<BoundedSource<T>> splitIntoBundles(
+      CassandraIO.Read<T> spec,
+      long desiredBundleSizeBytes,
+      SerializableFunction<PipelineOptions, CassandraService<T>> serviceFactory) {
+    long numSplits = 1;
+    List<BoundedSource<T>> sourceList = new ArrayList<>();
+    try {
+      if (desiredBundleSizeBytes > 0) {
+        numSplits = getEstimatedSizeBytes() / desiredBundleSizeBytes;
+      }
+    } catch (Exception e) {
+      // fallback to 1
+      numSplits = 1;
+    }
+    if (numSplits <= 0) {
+      numSplits = 1;
+    }
+
+    LOG.debug("Number of splits is {}", numSplits);
+    System.out.println("Number of splits: " + numSplits);
+
+    if (numSplits == 1) {
+      sourceList.add(new CassandraIO.CassandraSource<T>(spec, null, serviceFactory));
+      return sourceList;
+    }
+
+    Partitioner partitioner = new Partitioner(connectionConfiguration.getCluster());
+
+    BigInteger startRange = partitioner.minToken;
+    BigInteger endRange = partitioner.maxToken;
+    BigInteger startToken, endToken;
+
+    endToken = startRange;
+    BigInteger incrementValue =
+        (endRange.subtract(startRange)).divide(new BigInteger(String.valueOf(numSplits)));
+    String splitQuery;
+    for (int splitCount = 1; splitCount <= numSplits; splitCount++) {
+      startToken = endToken;
+      endToken = startToken.add(incrementValue);
```

Comment:
`endToken = (splitCount == numSplits) ? endRange : startToken.add(incrementValue)` - it's generally easier to read code when variables are not reassigned multiple times.